rfcs/0048-BYOC-Marvell-ML-accelerator-integration.md

```diff
@@ -0,0 +1,547 @@
+        GetExternalJsonWriter()->RegisterCB(this, &MrvlExtJson::GetExternalJSON);
+      }
+      virtual ~MrvlExtJson() {}
+      void GetExternalJSON(dmlc::JSONWriter* writer, Array<tvm::runtime::Module> external_mods,
+                           std::vector<GraphObjectPtr> nodes, std::vector<GraphNodeRef> heads);
+      void LoadExternalJsonAttrs(std::unordered_map<std::string, GraphAttrs>* external_attrs_map,
+                                 const Array<tvm::runtime::Module>& external_mods);
+    };
+```
+
+* the need to link between pre-trained model and final Marvell backend layer - for instance, through tvm_custom
+    * We did not include prototype code in PR-9730 but intend to provide our sample changes in another RFC and PR.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* We haven't identified any major *not* do items. Several other designs are by choices - that is we understand that
+  there are benefits for doing or benefits for not-doing.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* We follow the TVM BYOC framework to enable BYOC Marvell flow without impacting any TVM core features.
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* We are following the existing TVM BYOC framework and example files.
+    * for example: to do IR compositions, to define own IR passes, to mix implementations in Python/C++, and etc.
+
+* We have extended graph_executor_codegen.cc and JSON loader/saver in order to read and write out Marvell specific
+  attributes
+
+* Currently, we haven't spend enough time to under how tvm/rust/cargo requirements and steps. Therefore, we are
+  bypassing the tvm/Jenkinsfile's tests/scripts/task_rust.sh step. We will need help to re-enable the step.
+
+* We like to duplicate the Jenkins environment in order to run tvm/Jenkinsfile as is, but, we ran into many issues.
+  Currently, we have a tvm-like Jenksinsfile environment to only run a subset of test suites using a modified
+  Jenkinsfile.
+
+* We have identified a need to allow a call-back function to be registered when generating Mrvl-BYOC-specific
+  Nodes-JSON file. We are trying to follow TVM Python/CPP-CB style as much as possible. But, since our callback
+  function tvm/src/relay/backend/contrib/mrvl/graph_executor_codegen_mrvl.cc::GetExternalJSON() function is using
+  non-simple argument types, we need help from TVM community to provide suggestions/guidelines in order to make
+  new CB code better to meet TVM community requirements here.
+
+* For one Mrvl-BYOC relay transformation pass, we have identified a need to inject a (global) expr node ID for the
+  RelayExprNode class and its derived classes: Tuple and CallNode, so that during the transformation pass, we can
+  uniquely identify each Tuple or CallNode object. Again, we need help from TVM community to provide
+  suggestions/guidelines here in order to know whether this is one of the best ways to achieve the Mrvl-BYOC need.
```

Comment:
i think it would help to spell out why you guys need to be able to identify each expression here.

Reply:
yes but not just us but a data-scientist customer who is using TVM flow may like to know, for example, the linkages between the runtime performance #s (which were provided by driver and/or hardware) and their corresponding user frontend model’s operators (e.g., each expression which customer knows)

Reply:
yes and I have update the RFC to include more information regarding exprnode_id and its usages.